akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -548,3 +545,210 @@ final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]
+        close(resource)
+        resource = open()
+      }
+
+      private def closeStage(): Unit =
+        try {
+          close(resource)
+          doCompleteStage()
+        } catch {
+          case NonFatal(ex) ⇒ doFailStage(ex)
+        }
+
+      private def doFailStage(th: Throwable): Unit = {
+        completion.failure(th)
+        failStage(th)
+      }
+
+      private def doCompleteStage(): Unit = {
+        completion.success(Done)
+        completeStage()
+      }
+    }
+
+    (logic, completion.future)
+  }
+
+  override def toString = "FoldResourceSink"
+}
+
+/**
+ * INTERNAL API
+ */
+final class FoldResourceSinkAsync[T, S](
+  open:      () ⇒ Future[S],
+  writeData: (S, T) ⇒ Future[Unit],
+  close:     (S) ⇒ Future[Unit])
+  extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.out")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val completion = Promise[Done]()
+    val logic: GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var resource = Promise[S]()
+      implicit val context = ExecutionContexts.sameThreadExecutionContext
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        createStream()
+        setKeepGoing(true)
+      }
+
+      private def createStream(): Unit = {
+        val cb = getAsyncCallback[Try[S]] {
+          case scala.util.Success(res) ⇒
+            resource.success(res)
+            if (!isClosed(in)) {
+              pull(in)
+            }
+          case scala.util.Failure(t) ⇒ doFailStage(t)
+        }
+        try {
+          open().onComplete(cb.invoke)
+        } catch {
+          case NonFatal(ex) ⇒ doFailStage(ex)
+        }
+      }
+
+      private def onResourceReady(f: (S) ⇒ Unit): Unit = {
+        val cb = getAsyncCallback[S](f)
+        resource.future.onSuccess {
+          case r ⇒ cb.invoke(r)
+        }
+      }
+
+      val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop    ⇒ closeAndThen(() ⇒ doFailStage(ex))
+          case Supervision.Restart ⇒ restartState()
+          case Supervision.Resume ⇒
+            if (!isClosed(in)) {
+              pull(in)
+            }
+        }
+      }
+      val callback = getAsyncCallback[Try[Unit]] {
+        case scala.util.Success(_) ⇒
+          if (!isClosed(in)) {
+            pull(in)
+          }
+        case scala.util.Failure(t) ⇒ errorHandler(t)
+      }.invoke _
+
+      final override def onPush(): Unit = onResourceReady {
+        r ⇒
+          val elem = grab(in)
+          try {
+            writeData(r, elem).onComplete(callback)
```

Comment:
If it is already completed, we can avoid the callback and call that logic right away. Take a look at `MapAsync` for an example. We also know that callback is cheap, so it can be run on the calling thread, which is also done in `MapAsync`